### PR TITLE
Adding documentation for each API function in the bindings

### LINKED ISSUE #5: Add documentation for each API function
Each API function in these bindings needs to have documentation which gives information regarding the use of the function, the input arguments and output if there is a returning entity. As an example the python bindings use `docstring` style documentation blocks for each API function, check the [initialize_data](https://github.com/precice/python-bindings/blob/36e4782b8460ba93ec6bfc307bef7bf281c540fa/cyprecice/cyprecice.pyx#L115-L136) function for the documentation block. Two tasks would be involved here:
* Finding out if there is a standard way to write documentation for Julia function and in general Julia code
* Writing documentation for all API functions

### PR DESCRIPTION
**Pull request by pavelkharitenko**
Closes #5 

### The current documentation solution: 
1. Annotate every API function in `src/documentation.jl` and also put this documentation only on **precice.org**, no further packages or hosting on extra websites.

### Another possible solution: the default Julia Package way: 
1. We use [Documenter.jl](https://juliadocs.github.io/Documenter.jl/stable/) to generate Markdown files in a folder docs/.
2. We can host this documentation on its own website, it will look like [this example](https://evizero.github.io/Augmentor.jl/stable/)
3. We add on precice.org the link to this dedicated website that is hosting the documentation for PreCICE.jl.

To me, the second solution seems to be interessting only if the package has no project websites like **precice.org**, or is very complicated and requires its own website.
But maybe, the generated Markdown of Documenter.jl (they look like [this](https://github.com/Evizero/Augmentor.jl/blob/master/docs/src/index.md)) files might be compatible with the **precice.org**'s way of displaying markdown files.

### CODE REVIEW

**Comment by IshaanDesai** [comment]
@uekerman we have two ways of providing documentation, both of which have been mentioned above. What do you think of both ways? Which would be better?

**Comment by uekerman** [comment]
As discussed: option 1 (the current one) is sufficient.

**Comment by IshaanDesai** [comment]
@erikscheurer can you please resolve the merge conflicts this branch has? That would make reviewing easier for me. Just ask me if something is not clear in the conflict resolution.

File: src/PreCICE.jl [hunk 1]
```diff
@@ -1,11 +1,17 @@
 module PreCICE
 """
 The `PreCICE` module provides the bindings for using the preCICE api. For more information, visit https://precice.org/.
-"""
+"""
```
Comment (IshaanDesai) [inline src/PreCICE.jl:4]: Is this unintentional?

File: src/PreCICE.jl [hunk 2]
```diff
@@ -1,11 +1,17 @@
 module PreCICE
 """
 The `PreCICE` module provides the bindings for using the preCICE api. For more information, visit https://precice.org/.
-"""
+""" 
 
 
-# TODO add 'return nothing' keyword to void functions
+# These were proposed by the previous maintainer:
+# TODO add 'return nothing' keyword to void functions    # does this make sense? not done by documenter.jl
```
Comment (IshaanDesai) [inline src/PreCICE.jl:8]: Lets remove this if not found relevant

File: src/PreCICE.jl [hunk 3]
```diff
@@ -33,8 +39,23 @@ export
     getVersionInformation, actionWriteInitialData, actionWriteIterationCheckpoint, actionReadIterationCheckpoint
 
 
+@doc """
+
+    createSolverInterface(participantName::String, configFilename::String, solverProcessIndex::Integer, solverProcessSize::Integer)
```
Comment (IshaanDesai) [inline src/PreCICE.jl:44]: Is there a reason why there is indentation here and not in the documentation lines below? Is this how it is done anyway?

File: src/PreCICE.jl [hunk 4]
```diff
@@ -64,227 +85,1031 @@ function createSolverInterfaceWithCommunicator(participantName::String,
...
+
+    initialize()::Float64
+
+Fully initializes preCICE.
+This function handles:
+- Parallel communication to the coupling partner/s is setup.
+- Meshes are exchanged between coupling partners and the parallel partitions are created.
+- **Serial Coupling Scheme:** If the solver is not starting the simulation, coupling data is received from the coupling partner's first computation.
+
+Return the maximum length of first timestep to be computed by the solver.
```
Comment (IshaanDesai) [inline src/PreCICE.jl:98]: Can we state the data type of the return argument just like we can do for input arguments? That would be helpful for the users

File: src/PreCICE.jl [hunk 5]
```diff
@@ -64,227 +85,1031 @@ function createSolverInterfaceWithCommunicator(participantName::String,
...
+    isCouplingOngoing()::Bool
+
+Check if the coupled simulation is still ongoing.
+A coupling is ongoing as long as
+ - the maximum number of timesteps has not been reached, and
+ - the final time has not been reached.
+
+The user should call [`finalize`](@ref) after this function returns false.
+
+
```
Comment (IshaanDesai) [inline src/PreCICE.jl:216]: Too many blank lines here?

File: src/PreCICE.jl [hunk 6]
```diff
@@ -64,227 +85,1031 @@ function createSolverInterfaceWithCommunicator(participantName::String,
...
+v1_id = setMeshVertex(mesh_id, [1,1,1])
+```
+"""
 function setMeshVertex(meshID::Integer, position::AbstractArray{Float64})
     id::Integer = ccall((:precicec_setMeshVertex, "libprecice"), Cint, (Cint, Ref{Float64}), meshID, position)
     return id
 end
 
 
+# TODO Example has wrong arguments
```
Comment (IshaanDesai) [inline src/PreCICE.jl:469]: Do you want some clarification here?

File: src/PreCICE.jl (hunk 1 above)
Comment (erikscheurer) [reply src/PreCICE.jl:4]: Yes, sorry :)

File: src/PreCICE.jl (hunk 2 above)
Comment (erikscheurer) [reply src/PreCICE.jl:8]: I just found the [convention](https://docs.julialang.org/en/v1/manual/functions/#Returning-nothing) to add `return nothing` at the end of each function, so this should probably be done.

File: src/PreCICE.jl (hunk 3 above)
Comment (erikscheurer) [reply src/PreCICE.jl:44]: Yes, in julia documentation the Signature of the function is indented to signal to julia that it is the signature and not some other text.

File: src/PreCICE.jl (hunk 4 above)
Comment (erikscheurer) [reply src/PreCICE.jl:98]: The return type is given by `initialize()::Float64` in the signature, do you think this is enough or should we also put this in the text?

File: src/PreCICE.jl (hunk 5 above)
Comment (erikscheurer) [reply src/PreCICE.jl:216]: Definitely

File: src/PreCICE.jl (hunk 6 above)
Comment (erikscheurer) [reply src/PreCICE.jl:469]: Yes, I'm not sure what is correct here. Does the c function return something or save the vertices in the `positions::AbstractArray{Float64}`? 
And should we also pass the array by reference or return the resulting `positions`?

File: src/PreCICE.jl [hunk 7]
```diff
@@ -64,227 +85,1031 @@ function createSolverInterfaceWithCommunicator(participantName::String,
...
+WARNING: This routine is supposed to be used, when no edge information is available per se.
+        Edges are created on the fly within preCICE. This routine is significantly slower than the one
+        using edge IDs, since it needs to check, whether an edge is created already or not.
+
+# Arguments
+- `meshID::Integer`: ID of the mesh to add the edge to.
+- `firstVertexID::Integer`: ID of the first vertex of the edge.
+- `secondVertexID::Integer`: ID of the second vertex of the edge.
+- `thirdEdgeID::Integer`: ID of the third edge of the triangle.
+
```
Comment (IshaanDesai) [inline src/PreCICE.jl:657]: Unnecessary blank line

File: src/PreCICE.jl [hunk 8]
```diff
@@ -64,227 +85,1031 @@ function createSolverInterfaceWithCommunicator(participantName::String,
...
+values = [v1_x, v1_y, v1_z; v2_x, v2_y, v2_z; v3_x, v3_y, v3_z; v4_x, v4_y, v4_z; v5_x, v5_y, v5_z]
+writeBlockVectorData(data_id, vertex_ids, values)
+```
+"""
 function writeBlockVectorData(dataID::Integer, size::Integer, valueIndices::AbstractArray{Cint}, values::AbstractArray{Float64})
     ccall((:precicec_writeBlockVectorData, "libprecice"), Cvoid, (Cint, Cint, Ref{Cint}, Ref{Cdouble}), dataID, size, valueIndices, values)
 end
 
 
+# TODO Are they provided as a block of continuous memory?
```
Comment (IshaanDesai) [inline src/PreCICE.jl:771]: I dont really understand what you mean here, can you please elaborate?

File: src/PreCICE.jl [hunk 9]
```diff
@@ -64,227 +85,1031 @@ function createSolverInterfaceWithCommunicator(participantName::String,
...
+- `values::AbstractArray{Float64}`: The array holding the values.
+    
+# Notes
+
+Previous calls:
+ - Count of available elements at `values` matches the given size
+ - Count of available elements at `vertex_ids` matches the given size
+ - [`initialize`](@ref) has been called
 
+# Examples
```
Comment (IshaanDesai) [inline src/PreCICE.jl:838]: Blank code line in comments

File: src/PreCICE.jl [hunk 10]
```diff
@@ -64,227 +85,1031 @@ function createSolverInterfaceWithCommunicator(participantName::String,
...
+vertex_ids = [1, 2, 3, 4, 5]
+values = [1, 2, 3, 4, 5]
+writeBlockScalarData(data_id, vertex_ids, values)
+```
+"""
 function writeBlockScalarData(dataID::Integer, size::Integer, valueIndices::AbstractArray{Cint}, values::AbstractArray{Float64})
     ccall((:precicec_writeBlockScalarData, "libprecice"), Cvoid, (Cint, Cint, Ref{Cint}, Ref{Cdouble}), dataID, size, valueIndices, values)
 end
 
+# TODO example is strange
```
Comment (IshaanDesai) [inline src/PreCICE.jl:852]: Example looks okay to me. You can also use some dummy value instead of `value = v5`, like `value = 1.0`

**Review by IshaanDesai — CHANGES_REQUESTED** [review]
Overall looks good, thank you for the effort :+1: 
Some minor comments from my side, lets discuss each comment here in the PR

File: src/PreCICE.jl (hunk 4 above)
Comment (IshaanDesai) [reply src/PreCICE.jl:98]: Okay, so the data type next to the function name is the return type. Got it, looks good, lets stay with this :+1:

File: src/PreCICE.jl (hunk 8 above)
Comment (erikscheurer) [reply src/PreCICE.jl:771]: In the python bindings it said that the Values would be provided as a block of continuous memory. Is this also true in this case or would we be on the safe side not including this part?

File: src/PreCICE.jl (hunk 2 above)
Comment (IshaanDesai) [reply src/PreCICE.jl:8]: Lets add this wherever relevant :+1:

File: src/PreCICE.jl (hunk 9 above)
Comment (erikscheurer) [reply src/PreCICE.jl:838]: What do you mean here?

Comment (IshaanDesai) [reply src/PreCICE.jl]: False alarm sorry :sweat_smile:

File: src/PreCICE.jl (hunk 6 above)
Comment (IshaanDesai) [reply src/PreCICE.jl:481]: The intended use of `getMeshVertices` is that the user will provide a mesh ID and several vertex IDs and the function will return the corresponding vertex coordinates. We should return the resulting `positions` and not the reference.

Comment (erikscheurer) [reply src/PreCICE.jl:481]: Okay so I should change the function to not take positions as an input argument but instead to return the result of the c function call?

Comment (IshaanDesai) [reply src/PreCICE.jl:481]: Yes that is the appropriate usage

**Comment by erikscheurer** [comment]
In general, should the user have to specify the size, aka the number of vertices like in [`setMeshVertices`](https://github.com/precice/julia-bindings/pull/7/files#diff-838c0011f43392e6ba9eb00528458794ed058a0b102ec78930362168a0a8dbf6L177) or should the size be extracted from the length of the given vectors? In [Python](https://github.com/precice/python-bindings/blob/dadad118702f56bf760d1cc4da277b3c7900f64d/cyprecice/cyprecice.pyx#L483), the size is not given as an argument.

**Comment by erikscheurer** [comment]
One more, and maybe bigger Question is the format of the Arrays used. At the moment, they have the same shape as in C++:
(d0x,d0y,d0z,d1x,d1y,...,dny,dnz)
But in Python the Array is reshaped and given as a 3xn matrix. Julia also has the functionality of Matrices, should the Arrays be reshaped in such a way?

**Comment by IshaanDesai** [comment]
> In general, should the user have to specify the size, aka the number of vertices like in [`setMeshVertices`](https://github.com/precice/julia-bindings/pull/7/files#diff-838c0011f43392e6ba9eb00528458794ed058a0b102ec78930362168a0a8dbf6L177) or should the size be extracted from the length of the given vectors? In [Python](https://github.com/precice/python-bindings/blob/dadad118702f56bf760d1cc4da277b3c7900f64d/cyprecice/cyprecice.pyx#L483), the size is not given as an argument.

If it is possible to extract the size from the array itself then yes we should extract this. Can you check if this is possible in Julia? In Python we can extract the size and so we do it. In C and C++ we cannot do it so we ask the user to pass the size.
For now lets pass the size and just investigate if extracting the size of an array is possible in Julia. If it is possible then lets handle this in a different issue and PR

**Pull request merged**